### ISSUE TITLE
webapp uid/gid changing on upgrade breaks application expectations

### ISSUE THREAD

**Issue by christopherowen** [aws/elastic-beanstalk-roadmap#137]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave "+1" or "me too" comments. They generate extra noise for issue followers and do not help prioritize the request.

**Tell us about your request**
On beanstalk update, it is possible for the uid and gid to change for the 'webapp' user.

This can break expectations for long term storage (such as for an EFS filesystem mount).

I would like to have the uid/gid for the webapp user be consistent or otherwise configurable (to be configured consistently).

If there is a reason that the uid/gid must change, I would like to know why and have a strategy suggested to identify and address it at the time of update.

**Is this request specific to an Elastic Beanstalk [platform](https://docs.aws.amazon.com/elasticbeanstalk/latest/platforms/platforms-supported.html)?**
If so, which one(s)?

I can confirm that this applies to, at least, "PHP 7.4 AL2 version 3.1.4"

**Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?**

My webapps require permanent filesystem storage.

We achieve this via efs mounted at /efs.

The file permissions are set such that the webapp users has the appropriate read and write permissions.
 
**Are you currently working around this issue?**

When beanstalk updates happen, the webapps break, and we have to investigate and chown the files to the new pid/gid.

As the beanstalks upgrade on their own schedules, it is very surprising when things break.

**Additional context**

This was additionally reported on the forum here:

https://forums.aws.amazon.com/thread.jspa?threadID=330170

**Comment by ian-a-anderson** [comment]
Is it possible (or advisable) to workaround this issue by changing the UID/GID of the webapp user on deployment, via an .ebextensions or .platform script? The following seems to work on my dev beanstalk environment, but I am unsure of the consequences of altering this user/group in this way. Could anyone from the EB team advise on whether this is a valid approach?

-------- example adjust-webapp-user.sh script --------

\# fail on error
set -e

\# Exit if user has correct id
if id -u webapp | grep -q '498'; then echo 'webapp user id correct'; exit 0; fi;

\# terminate php-fpm, as it spawns threads owned by webapp
killall -KILL php-fpm;

\# adjust the Elastic Beanstalk webapp user as AL2 has a diff ID
usermod -u 498 webapp;
groupmod -g 496 webapp;

\# CHOWN the downloaded deployment to match the webapp user
\# since those files were downloaded using the original UID/GID
sudo chown -R webapp:webapp /var/app

\# restart php-fpm
systemctl restart php-fpm.service;

\# output success
echo 'webapp user id adjusted';

**Comment by matthewdouglas** [comment]
I'm also using EFS and ran into this problem. I've been using a custom AMI that I build with EC2 Image Builder, so what I've done there is add a step to set the UID and GID as I expect when the AMI is baked. Have been doing this in a production environment for a while now and no issues so far.

**Comment by 1tonyca** [comment]
Please refer to release notes for details
https://docs.aws.amazon.com/elasticbeanstalk/latest/relnotes/release-2022-02-03-linux.html

**Comment by nodomain** [comment]
What is your recommended migration/mitigation strategy? This will presumably break a lot of our environments during scheduled upgrades.

**Comment by ian-a-anderson** [comment]
I second the question above question from @nodomain... we have hundreds of thousands of files on EFS and taking downtime to change ownership would be problematic. Given that, does AWS have any comment on the workaround script I posted above? Will this still work going forward with the new 900 ID? We have been running using this script for our affected environments and it seems to work as needed, but I am curious if there is some other recommended alternative.

**Comment by nickpearson** [comment]
I'm dealing with production downtime as a result of this right now. The [update documentation](https://docs.aws.amazon.com/elasticbeanstalk/latest/relnotes/release-2022-02-03-linux.html) specifies that the new UID and GID are set to 900 and 900 "for new environments", but it was applied to my existing environments that use weekly instance replacement, as @nodomain correctly presumes.

I have two EB environments that use EFS for shared storage between them (web and worker tiers), and it's configured with weekly auto-update and instance replacement. When the instances were replaced over the weekend, the `webapp` UID and GID changed from 995 and 993 (respectively) to 900 and 900. This means all of the EFS files and directories are currently inaccessible to the app, while I wait on a recursive `chown` to update them.

It would have been faster to change the `webapp` user's UID/GID, but I don't know if that would cause other problems, because 995 is in use as a group ID for the `ec2-instance-connect` user (found in `/etc/passwd`).

In case it helps someone, here's a prebuild platform script I use on a few other production EB environments with a similar setup. (I wrote the script a couple months ago using @ian-a-anderson's [above script](https://github.com/aws/elastic-beanstalk-roadmap/issues/137#issuecomment-821532968) as inspiration — thanks for that!). These environments also use weekly instance replacement, and when upgrading from an older EB platform version, I needed to access EFS from both the existing environment and the new one, so the UIDs/GIDs had to match. Thus, the new env uses this script to change the UID/GID to 496/494 like they originally were on the older Amazon Linux (1) instances and then `chmod`s all of the known `webapp`-owned files on the local filesystem to use the old UID/GID (with the end result being that the local files and EFS files use the same UID/GID, without me having to change all of the EFS files, which would have taken a very long time).

I think the better option is to update your EFS files to use the new 900/900 IDs documented in the platform update, but if you have too many files for that to be feasible on short notice, this may work for you. (This script saved me from the same production downtime on the environments of mine that use it.)

**IMPORTANT:** This changes the UID/GID for the `webapp` user to their **old** values. If you want to use this for your environment, set the `EC2_WEBAPP_UID` and `EC2_WEBAPP_GID` variables in the script to whatever your EFS (or whatever) currently uses. Also be sure your IDs don't conflict with other user/group IDs. (I don't know enough about it to know what kinds of problems that would cause, or whether the `usermod` commands would fail in that case.) Before using this script in a production environment, you should clone your EB environment and try it out there.

I have this script at `.platform/hooks/prebuild/01_ids.sh`. Hopefully this helps someone.

```bash
#!/bin/bash

# This script updates the numeric IDs of the webapp user and webapp group on
# Amazon Linux 2 (Elastic Beanstalk platform version 3) to match those on the
# original Amazon Linux (Elastic Beanstalk platform version 2) so we can
# read and write to the shared AWS Elastic File System (EFS) that was created
# using the original user and group IDs.

set -xe

EB_APP_USER=$(/opt/elasticbeanstalk/bin/get-config platformconfig -k AppUser)

EB2_WEBAPP_UID=496
EB2_WEBAPP_GID=494

EB3_WEBAPP_UID="$(id -u "$EB_APP_USER")" # 900 (was 995 before 2022-02-03) on an unmodified instance
EB3_WEBAPP_GID="$(id -g "$EB_APP_USER")" # 900 (was 993 before 2022-02-03) on an unmodified instance

# exit if the IDs are already correct
if [[ "$EB3_WEBAPP_UID" == "$EB2_WEBAPP_UID" ]] && \
   [[ "$EB3_WEBAPP_GID" == "$EB2_WEBAPP_GID" ]]; then
  exit
fi

# store original user ID and group ID
BASE_DIR=/root/eb-deploy
mkdir -p $BASE_DIR
echo "$EB_APP_USER uid: $EB3_WEBAPP_UID ($(date -u +"%Y-%m-%dT%H:%M:%SZ"))" >> "$BASE_DIR/original-ids"
echo "$EB_APP_USER gid: $EB3_WEBAPP_GID ($(date -u +"%Y-%m-%dT%H:%M:%SZ"))" >> "$BASE_DIR/original-ids"

# get PIDs of processes currently owned by webapp user
user_pids=()
while IFS='' read -r line; do
  user_pids+=("$line")
done < <(pgrep -u "$EB_APP_USER")

# forcibly kill processes owned by the webapp user (usermod will fail otherwise)
if (( ${#user_pids[@]} )); then kill -9 "${user_pids[@]}"; fi

# change the user ID from 900 (was 995) to 496
usermod --uid $EB2_WEBAPP_UID "$EB_APP_USER"

# change the group ID from 900 (was 993) to 494
groupmod --gid $EB2_WEBAPP_GID "$EB_APP_USER"

# update group ID and user ID for the bulk of the webapp-owned files and directories
chown -R $EB2_WEBAPP_UID:$EB2_WEBAPP_GID /var/app

# get all root directories that are not `/` or `/efs`
root_dirs="$(find / -xdev -maxdepth 1 -type d -not -path / -not -path /dev -not -path /efs -not -path /proc)"

# update group ID and user ID of all matching files and directories
while read -r d; do
  find "$d" -xdev -not -path /var/app/\* -gid "$EB3_WEBAPP_GID" -exec chgrp -h $EB2_WEBAPP_GID {} \;
  find "$d" -xdev -not -path /var/app/\* -uid "$EB3_WEBAPP_UID" -exec chown -h $EB2_WEBAPP_UID {} \;
done <<< "$root_dirs"
```

**Comment by nodomain** [comment]
Thanks for the nice script. It would still be nice to get some official statement from AWS on this. I am trying this via our TAM as well.

Sometimes I cannot understand the reasoning behind things like this...

**Comment by kevinludlow** [comment]
@nickpearson : Firstly, thanks for adding that script in there.  It's a great starting resource if nothing else.

Are you able to briefly describe the environment you were running that in.  We have an EBS setup, Linux2, basic LAMP env, using the Application LB, but I when I tried deploying your script, it failed every time because it wasn't able to shut down all of the processes owned by webapp.

Of course I looked through the script and realized that what was happening was that the php-fpm.service would spin up a new process the moment your script shut them down.

So...I modified the script slightly to stop the php-fpm.service just before the script starts.  Let it do its thing.  And then start it back up.

This worked - kindof.  Our G/UIDs were changed to the old numbers that we needed, but it wound up breaking a ton of other services on the machine.  Ultimately Beanstalk wouldn't accept it and trashed the machine.  So I haven't been able to figure it out beyond that.

Anyway, I was just curious if perhaps you were using a radically different environment and perhaps there was something else I was missing.  We'd LIKE to use your solution, but I just haven't been able to get it to work yet.  It executes, but it doesn't seem to work quite right.

Thanks in advance for any help.
